dvc/tree/gdrive.py

```diff
@@ -510,35 +511,71 @@ def exists(self, path_info, use_dvcignore=True):
         else:
             return True
 
-    def _list_paths(self, prefix=None):
-        if not self._ids_cache["ids"]:
-            return
+    def ls(
+        self, path_info, detail=False, recursive=False
+    ):  # pylint: disable=arguments-differ
+        if recursive:
+            for entry in self.walk_files(path_info, detail=detail):
+                if detail:
+                    yield entry
+                else:
+                    yield entry.path
+            return None
 
-        if prefix:
-            dir_ids = self._ids_cache["dirs"].get(prefix[:2])
-            if not dir_ids:
-                return
+        cached = path_info.path in self._ids_cache["dirs"]
+        if cached:
+            dir_ids = self._ids_cache["dirs"][path_info.path]
         else:
-            dir_ids = self._ids_cache["ids"]
-        parents_query = " or ".join(
-            f"'{dir_id}' in parents" for dir_id in dir_ids
-        )
-        query = f"({parents_query}) and trashed=false"
+            dir_ids = self._path_to_item_ids(path_info.path)
+
+        if not dir_ids:
+            return None
 
+        query = " or ".join(f"'{dir_id}' in parents" for dir_id in dir_ids)
+        query = f"({query}) and trashed=false"
+
+        root_path = path_info.path
         for item in self._gdrive_list(query):
-            parent_id = item["parents"][0]["id"]
-            yield posixpath.join(
-                self._ids_cache["ids"][parent_id], item["title"]
-            )
+            item_path = posixpath.join(root_path, item["title"])
+            if detail:
+                if item["mimeType"] == FOLDER_MIME_TYPE:
+                    yield {"type": "directory", "name": item_path}
+                else:
+                    yield {
+                        "type": "file",
+                        "name": item_path,
+                        "size": item["fileSize"],
+                        "checksum": item["md5Checksum"],
+                    }
+            else:
+                yield item_path
+
+        if not cached:
+            self._cache_path_id(root_path, min(dir_ids))
+
+    def walk(
+        self, path_info, prefix=None, detail=False
+    ):  # pylint: disable=unused-argument
+        stack = deque([path_info])
+        while stack:
+            path_info = stack.pop()
+            files, directories = [], []
+            for entry in self.ls(path_info, detail=True):
+                path = path_info.replace(path=entry["name"])
+                if entry["type"] == "file":
+                    if detail:
+                        files.append(entry)
+                    else:
+                        files.append(path)
+                else:
+                    directories.append(path)
+
+            yield path_info, directories, files
+            stack.extendleft(directories)
 
     def walk_files(self, path_info, **kwargs):
```

Comment:
just for the record and so that we are on the same page ... we are making an extra call here now to list `00`, `01`, ... `ff` directories first, right? Before we were directly taking `dir_ids` from the cache (since we expected cache to contain only storage related ids in it)

Reply:
> we are making an extra call here now to list 00, 01, ... ff directories first, right?

yes. We actually have the ids, so we don't make any extra calls to list the directory names themselves but we make extra calls to list their contents. So if you run `dvc gc -c ` twice and if some directories are empty, previous implementation would result within a single call (since it doesn't actually have to take responsibility of other stuff, and works recursively) but now it makes additional calls for contents of each directory which is a bit slower (but that is how fsspec does it too). CC: @efiop , we had some internal discussion with Ruslan (in case of I am missing something)

Reply:
okay, so it's even 1 vs 256+ sequential requests, correct? `gc` is fine, I'm worried a bit that this is also part of the regular `push` flow as far as I remember.

Can it be/should it be part of the `ls(recursive=True)` though?

It's a bit strange btw that ls depends on walk_files and walk_file depends on ls. Feels like `ls` should be self contained. And it can be optimized when `recursive=True` to batch/parallelize requests. Probably it won't be recursive itself then.

Reply:
> Can it be/should it be part of the ls(recursive=True) though?

As far as I can tell, not really. At least not without knowing the whole structure of the tree (not just the initials levels), we can initially go with the first one and then visit the remaining parts though that feels like very cumbersome logic so I am not sure about it. 

I might be wrong, though it seems like there is no change in the push stuff? (have to re-check tomorrow)

Reply:
> I might be wrong, though it seems like there is no change in the push stuff? (have to re-check tomorrow)

it depends on how `push` decides to do the check for the existing files. It kinda runs status before. (as far as I remember)

> As far as I can tell, not really.

could you elaborate please?

>  At least not without knowing the whole structure of the tree (not just the initials levels), we can initially go with the first one and then visit the remaining parts though that feels like very cumbersome logic so I am not sure about it.

you just accumulate directories (similar to your deque) and batch them when you do a call. Yes, it will require some additional logic to accumulate directories within the ls itself and batch them. Doesn't sound very terrible to me.

(again, discuss with @efiop the priorities here, I'm commenting and reviewing since I have a bit of experience with GD and understand some implications. It might be fine to create tickets for some things. It's great to have a plan at least if we are migrating to the new interface).

Reply:
> you just accumulate directories (similar to your deque) and batch them when you do a call. Yes, it will require some additional logic to accumulate directories within the ls itself and batch them. Doesn't sound very terrible to me.

Sure, that's what I meant. Though as I stated earlier, this would be very local and unfortunately won't be available on other solutions in the future (like gdrivefs). Will try to work on this tomorrow though 👀

Reply:
I have an implementation ready, which should work much efficiently though it is only acceptable if we refactor `.walk_files()` to use `.ls(recursive=True)` instead of `.walk()`. By doing that, we can also remove the `.walk()` method for now since there won't be any use case for it. Would it be a problem to disregard fsspec's way on this @efiop?

Reply:
@isidentical Doesn't seem like that would be a problem. If we don't need walk - let's get rid of it :slightly_smiling_face: